Corvus.Extensions.Samples/TraversalExtensionsSample.dib

```diff
@@ -0,0 +1,242 @@
+#!markdown
+
+## Traversal Extensions - Sample Notebook
+
+#!markdown
+
+First, add a reference to the `Corvus.Extensions` NuGet package.
+
+#!csharp
+
+#r "nuget: Corvus.Extensions, 1.1.4"
+
+using Corvus.Extensions;
+
+#!markdown
+
+### ForEachAsync()
+
+#!markdown
+
+Execute an async action for each item in the enumerable.
+
+#!csharp
+
+public async Task IsEqualToOne(int x)
+{
+    bool result = x.Equals(1);
+    Console.WriteLine(result);
+}
+
+#!csharp
+
+var list = new List<int> { 1, 2, 3, 4, 5 };
+await list.ForEachAsync<int>(IsEqualToOne)
+
+#!markdown
+
+The `ForEachAsync()` method is here for historical interest. It is now possible to write the following and get the same output:
+
+#!csharp
+
+foreach (int x in list)
+{
+    await IsEqualToOne(x);
+}
+
+#!markdown
+
+### ForEachAtIndex<T>()
+
+#!markdown
+
+Execute an action for each item in the enumerable with the index of the item in the enumerable.
+
+#!csharp
+
+static void ConsolePrint(string s, int i)
+{
+    Console.WriteLine(i);
+    Console.WriteLine(s);
+}
+
+Action<string, int> printAction = ConsolePrint;
+
+#!csharp
+
+IEnumerable<string> list = new List<string> { "foo", "bar", "quux" };
+list.ForEachAtIndex<string>(printAction);
+
+#!csharp
+
+int i = 0;
+
+foreach (string v in list)
+{
+    printAction(v, i++);
+}
+
+#!markdown
+
+### ForEachAtIndexAsync()
+
+#!markdown
+
+Execute an async action for each item in the enumerable, in turn, with the index of the item in the enumerable.
+
+#!csharp
+
+static async Task ConsolePrint(string s, int i)
+{
+    Console.WriteLine(s);
+    Console.WriteLine(i);
+}
+
+#!csharp
+
+IEnumerable<string> list = new List<string> { "foo", "bar", "quux" };
+await list.ForEachAtIndexAsync<string>(ConsolePrint);
+
+#!markdown
+
+### ForEachFailEnd()
+
+#!markdown
+
+Execute an action for each item in the enumerable. 
+
+If any operation fails, then the enumeration is continued to the end when an Aggregate Exception is thrown containing the exceptions thrown by any failed operations.
```

Comment:
Consider adding a sentence describing a scenario in which this is useful. E.g., we use exactly this sort of thing in some unit tests to clean up things that were set up for the test. We do it to make sure that all necessary cleanup is attempted, even if individual steps fail.

(It's especially important with tests that create things in Azure that cost money. When tests complete, you want to make sure that if one tear-down step happens to fail, you don't just give up immediately, failing to delete potentially expensive resources.)

Reply:
I have added the following explanation:

> This is useful when cleaning up Azure resources that were set up for testing purposes, for example. It makes sure that even if one step fails, the process doesn't stop, as this would mean that some potentially expensive resources aren't deleted.